# __Introduction to finpy-tse Python Module__
<font color='green'>__A Free Python Module to Access Tehran Stock Exchange Historical and Real-Time Data__</font>
<br> <font color='green'>__A. RAHIMI__</font>


#### __Motivation:__ 

When I started learning quantitative portifolio management approaches and algorithmic trading models, my main barrier in going forward and examining the learned ideas was hardship in accessing the data. I had the tools for exploring the data but not the data itself! This was my main motivation for developing this module, maybe I would be able to help in removing this barrier from new learners's path, so that, they could focus on the data analysis and machine learning tools and do not worry about the data.

#### __Scope:__  
    
The scope of this module is just __getting the data into your python enviroment__. You can build you own databases, if you are interested, using these data and manage it in appropriate way to satisfy your purposes. Using this module, you will have most of the data, you may need for your research or development purposes. The finpy-tse module will provide you the historical and real-time data of the Tehran Stock Exchange and I am pretty sure that you can leverage module-provided data to build any other specific form of price-related data, for your project, either educational or scientific, and even commercial! 

In this notebook, we will explain finpy-tse module's function and you will learn how to use it to access to your desired data. Besides, we have a __Farsi Handbook__ for those who may not know English enough, to understand what we will bring in this notebook. You can find that handbook and some video tutorials in our Github page, YouTube and Telegram channel.
    
#### __Acknowledgment:__

I also would like to thank __Mr. Rasool Gafoori__ and __Ms. Marziyeh Nourahmadi__ for their contribution in module development and preparing its Farsi handbook.

In [4]:
import finpy_tse as fpy
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


---------------------------------------------
## __1. Reading Single Stock Data__

Most people do their analysis based on stock daily price data. In this section, we will introduce finpy-tse module functions, you can use to access Tehran Stock Exchange (TSE) stocks data including:

### __1.1. Get_Price_History() Function__
<font color='green'>__Access to Stock Daily Historical Price Data (= سابقه قیمت سهم)__</font>

This function reads a given stock price history for a given period of time starting from start_date till end_date (included). The out is a Pandas' dataframe with open, high, low, close (OHLC) and final prices. It also includes trading volume and money-values (in Rials) and number of trades, executed in a specific day. for stock, you can enter ticker or full name of the company, you are interested in. 


In [3]:
df_price_history = fpy.Get_Price_History(stock = 'ذوب', start_date = '1399-12-15', end_date='1401-01-01',\
                                         ignore_date = True, adjust_price = False, show_weekday = True, double_date = True)
df_price_history

Date    Weekday  Open  High   Low  Close  Final     Volume  \
J-Date                                                                        
1390-11-19 2012-02-08  Wednesday  1304  1304  1304   1304   1304  248387500   
1390-11-23 2012-02-12     Sunday  1369  1369  1320   1369   1365   75848282   
1390-11-24 2012-02-13     Monday  1379  1397  1301   1311   1350   23435322   
1390-11-25 2012-02-14    Tuesday  1320  1324  1293   1317   1304   12352753   
1390-11-26 2012-02-15  Wednesday  1311  1324  1282   1290   1301    9402675   
...               ...        ...   ...   ...   ...    ...    ...        ...   
1400-09-20 2021-12-11   Saturday  3530  3540  3401   3410   3463   32001668   
1400-10-05 2021-12-26     Sunday  3624  3675  3590   3605   3618   74715054   
1400-10-06 2021-12-27     Monday  3645  3798  3610   3798   3781  220940055   
1400-10-07 2021-12-28    Tuesday  3790  3845  3710   3770   3788   70894531   
1400-10-08 2021-12-29  Wednesday  3792  3840  3651   3661   3737   51089874   

                   Value    No Ticker                   Name  Part  
J-Date                                                              
1390-11-19  323897300000  2061    ذوب  سهامی ذوب آهن  اصفهان     2  
1390-11-23  103516867261  4290    ذوب  سهامی ذوب آهن  اصفهان     2  
1390-11-24   31646815791  2189    ذوب  سهامی ذوب آهن  اصفهان     2  
1390-11-25   16101932134  1051    ذوب  سهامی ذوب آهن  اصفهان     2  
1390-11-26   12234238887   680    ذوب  سهامی ذوب آهن  اصفهان     2  
...                  ...   ...    ...                    ...   ...  
1400-09-20  110826996114  2576    ذوب  سهامی ذوب آهن  اصفهان     2  
1400-10-05  270344068711  2090    ذوب         ذوب آهن اصفهان     1  
1400-10-06  835269163218  6849    ذوب         ذوب آهن اصفهان     1  
1400-10-07  268555627554  4236    ذوب         ذوب آهن اصفهان     1  
1400-10-08  190918050184  3307    ذوب         ذوب آهن اصفهان     1  

[2121 rows x 13 columns]

In [5]:
df_price_history['1395-02-04':]

Date    Weekday  Open  High   Low  Close  Final     Volume  \
J-Date                                                                        
1395-02-04 2016-04-23   Saturday  2629  2655  2531   2542   2586   12404982   
1395-02-05 2016-04-24     Sunday  2501  2648  2500   2621   2568   18554260   
1395-02-06 2016-04-25     Monday  2650  2696  2627   2670   2689   37686390   
1395-02-07 2016-04-26    Tuesday  2698  2823  2644   2823   2803   73721764   
1395-02-08 2016-04-27  Wednesday  2850  2888  2731   2760   2820   46757531   
...               ...        ...   ...   ...   ...    ...    ...        ...   
1400-09-20 2021-12-11   Saturday  3530  3540  3401   3410   3463   32001668   
1400-10-05 2021-12-26     Sunday  3624  3675  3590   3605   3618   74715054   
1400-10-06 2021-12-27     Monday  3645  3798  3610   3798   3781  220940055   
1400-10-07 2021-12-28    Tuesday  3790  3845  3710   3770   3788   70894531   
1400-10-08 2021-12-29  Wednesday  3792  3840  3651   3661   3737   51089874   

                   Value    No Ticker                   Name  Part  
J-Date                                                              
1395-02-04   32076733071  1538    ذوب  سهامی ذوب آهن  اصفهان     2  
1395-02-05   47652232835  2091    ذوب  سهامی ذوب آهن  اصفهان     2  
1395-02-06  101327597428  2599    ذوب  سهامی ذوب آهن  اصفهان     2  
1395-02-07  206678448532  4599    ذوب  سهامی ذوب آهن  اصفهان     2  
1395-02-08  131845081023  4181    ذوب  سهامی ذوب آهن  اصفهان     2  
...                  ...   ...    ...                    ...   ...  
1400-09-20  110826996114  2576    ذوب  سهامی ذوب آهن  اصفهان     2  
1400-10-05  270344068711  2090    ذوب         ذوب آهن اصفهان     1  
1400-10-06  835269163218  6849    ذوب         ذوب آهن اصفهان     1  
1400-10-07  268555627554  4236    ذوب         ذوب آهن اصفهان     1  
1400-10-08  190918050184  3307    ذوب         ذوب آهن اصفهان     1  

[1238 rows x 13 columns]

### __1.2. Get_RI_History() Function__
<font color='green'>__Access to Stock Historical Retail-Institutional Traders Data (= سابقه حقیقی-حقوقی سهم)__</font>

Here, R stands for retail traders and I stands for institutional traders. This function returns retail and institutional traders trades statistics in daily time frame, for a given stock, in a given period between start and end date.

| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

Returned dataframe's columns:

| __Column Name__ | __Description (EN)__ | __Description (FA)__ | 
| :- | :- | -: | 
| __J-Date__ | Jalali date | تاریخ شمسی |
| __Date__ | International date | تاریخ میلادی |
| __Weekday__ | Name of weekday | روز هفته |
| __No_Buy_R__ | Number of buy trades by retail traders | تعداد خرید حقیقی |
| __No_Buy_I__ | Number of buy trades by institutional traders | تعداد خرید حقوقی |
| __No_Sell_R__ | Number of sell trades by retail traders | تعداد فروش حقیقی |
| __No_Sell_I__ | Number of sell trades by institutional traders | تعداد فروش حقوقی |
| __Vol_Buy_R__ | Volume of buy trades by retail traders | حجم خرید حقیقی |
| __Vol_Buy_I__ | Volume of buy trades by institutional traders | حجم خرید حقوقی |
| __Vol_Sell_R__ | Volume of sell trades by retail traders | حجم فروش حقیقی |
| __Vol_Sell_I__ | Volume of sell trades by institutional traders | حجم فروش حقوقی |
| __Val_Buy_R__ | Value of buy trades by retail traders | ارزش خرید حقیقی |
| __Val_Buy_I__ | Value of buy trades by institutional traders | ارزش خرید حقوقی |
| __Val_Sell_R__ | Value of sell trades by retail traders | ارزش فروش حقیقی |
| __Val_Sell_I__ | Value of sell trades by institutional traders | ارزش فروش حقوقی |

In [8]:
df_ri_history = fpy.Get_RI_History(stock = 'پترول', start_date = '1395-01-01', end_date='1400-12-29',\
                                   ignore_date = True, show_weekday = False, double_date = False)
df_ri_history

No_Buy_R  No_Buy_I  No_Sell_R  No_Sell_I  Vol_Buy_R  Vol_Buy_I  \
J-Date                                                                       
1390-10-19       196         4         41          2    6541842    1154158   
1390-10-20         2         0          3          0     162000          0   
1390-10-21         2         0          6          0      36200          0   
1390-10-25         2         1          2          0      25000      10000   
1390-10-26        48         4         71          0    1544016     415842   
...              ...       ...        ...        ...        ...        ...   
1400-10-04      1881         4       1525          6   93885417    3939761   
1400-10-05      1431         6       1143          3   56021551    5490000   
1400-10-06       982         3       1299          2   36918673   13680000   
1400-10-07       929         5       1078          3   44457834    5817475   
1400-10-08      1094         2       1008          4   44021332    1943120   

            Vol_Sell_R  Vol_Sell_I     Val_Buy_R    Val_Buy_I    Val_Sell_R  \
J-Date                                                                        
1390-10-19     2496000     5200000    9283101039   1674237000    3716786019   
1390-10-20      162000           0     242190000            0     242190000   
1390-10-21       36200           0      56797800            0      56797800   
1390-10-25       35000           0      41175000     16470000      57645000   
1390-10-26     1959858           0    2661805966    718990818    3380796784   
...                ...         ...           ...          ...           ...   
1400-10-04    96089655     1735523  238650922338   9945866586  244216542053   
1400-10-05    53470781     8040770  137605408676  13498537668  131432981736   
1400-10-06    50456596      142077   92518320652  34493358301  126656975686   
1400-10-07    48619561     1655748  111873339285  14532090369  122209150333   
1400-10-08    43951820     2012632  108943621288   4762140578  108680431053   

             Val_Sell_I Ticker                         Name Part  
J-Date                                                            
1390-10-19   7240552020  پترول    گروه پتروشیمی س. ایرانیان    3  
1390-10-20            0  پترول    گروه پتروشیمی س. ایرانیان    3  
1390-10-21            0  پترول    گروه پتروشیمی س. ایرانیان    3  
1390-10-25            0  پترول    گروه پتروشیمی س. ایرانیان    3  
1390-10-26            0  پترول    گروه پتروشیمی س. ایرانیان    3  
...                 ...    ...                          ...  ...  
1400-10-04   4380246871  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1400-10-05  19670964608  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1400-10-06    354703267  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1400-10-07   4196279321  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  
1400-10-08   5025330813  پترول  گ.س.وت.ص.پتروشیمی خلیج فارس    1  

[2010 rows x 15 columns]

---------------------------------------------
## __2. Reading Single Stock Intraday Data__

Most people do their analysis based on stock daily price data. In this section, we will introduce finpy-tse module functions, you can use to access Tehran Stock Exchange (TSE) stocks data including:

### __2.1. Get_IntradayTrades_History() Function__

Tehran Stock Exchange id open from 8:45 to 12:30. Orders are executed from 9:00 to 12:30 and from 8:45 to 9:00 is pre-opening period. In pre-opening period, you can place orders but they are not executed until 9:00. This function returns every signle executed trade with its execution time, volume and price. You can use this type of output to create price bars in different time frames, tick charts, volume bars, information bars, etc. as mentioned is Marcos Lopez Deprada's book, i.e. "Advances in Financial Machine Learning". 

In [35]:
df_intraday = fpy.Get_IntradayTrades_History(stock = 'ذوب', start_date = '1400-09-20', end_date='1400-10-20',
                                             jalali_date = True, combined_datatime = False, show_progress = True)
df_intraday

Progress :  100.0 %


Volume  Price
J-Date     Time                   
1400-09-20 09:00:38   35699   3530
           09:00:38   32000   3530
           09:00:38   22301   3530
           09:00:38   20000   3530
           09:00:38   10964   3530
...                     ...    ...
1400-10-11 12:29:52   15000   3640
           12:29:56   14000   3639
           12:29:57   15000   3640
           12:29:59   12154   3640
           12:29:59   17517   3640

[21560 rows x 2 columns]

### __2.2. Get_IntradayLOB_History() Function__
<font color='green'>__Intra-day, Limit Order Book (LOB), Historical Data for a Single Stock__</font>

Tehran Stock Exchange has a queue-based limit order book (LOB). Five first rows of this LOB is accessible for public and this function gets that data for previous days. Remember, this is not real time data, but it is for the past days. You can use this data to create your own dataset and develop trading dtrategies based on what happens in supply and demand.

In [51]:
df_lob = fpy.Get_IntradayLOB_History(stock = 'ذوب', start_date = '1400-09-20', end_date='1400-10-20',
                                     jalali_date = True, combined_datatime = False, show_progress = True)
df_lob

Progress :  100.0 %


Sell_No Sell_Vol  Sell_Price  Buy_Price  Buy_Vol  \
J-Date     Time     Depth                                                    
1400-09-20 08:45:05 1           1     1450      3549.0     3549.0     2300   
                    2           1    10000      3580.0     3420.0     2913   
                    3           1     1600      3590.0     3410.0    57843   
                    4           1     2500      3598.0     3400.0     1750   
                    5           2    11700      3600.0     3380.0     6000   
...                           ...      ...         ...        ...      ...   
1400-10-11 12:29:52 1           5   252154      3640.0     3639.0  2169810   
           12:29:56 1           4   238154      3640.0     3639.0  2155810   
           12:29:57 1           4   223154      3640.0     3639.0  2140810   
                    2           1     2500      3641.0     3631.0   221490   
           12:29:59 1           3   193483      3640.0     3639.0  2140810   

                          Buy_No  Day_LL  Day_UL  
J-Date     Time     Depth                         
1400-09-20 08:45:05 1          1  3321.0  3669.0  
                    2          1  3321.0  3669.0  
                    3          3  3321.0  3669.0  
                    4          1  3321.0  3669.0  
                    5          1  3321.0  3669.0  
...                          ...     ...     ...  
1400-10-11 12:29:52 1         10  3551.0  3923.0  
           12:29:56 1         10  3551.0  3923.0  
           12:29:57 1          9  3551.0  3923.0  
                    2          3  3551.0  3923.0  
           12:29:59 1          9  3551.0  3923.0  

[37022 rows x 8 columns]

### __2.3. Get_Queue_History() Function__
<font color='green'>__Market Closing Queue History for a Given Stock and Period__</font>
<br><font color='gray'>__REVISION DATE: 1400-04-02__</font>

Tehran Stock Exchange has a queue-based limit order book (LOB). Five first rows of this LOB is accessible for public and this function gets that data for previous days. Remember, this is not real time data, but it is for the past days. You can use this data to create your own dataset and develop trading dtrategies based on what happens in supply and demand.


| __Parameter__ | __Description__ | 
| :- | :- | 
| __stock__ | company's stock __ticker__ or __full name__ in __Farsi__|
| __start_date__ | returned data start date in __Jalali__ and __YYYY-MM-DD__ format (inculded, if it is trading day) |
| __end_date__ | returned data end date in __Jalali__ and __YYYY-MM-DD__ format (inculded, if it is trading day) |

In [50]:
df_bq_sq_val = fpy.Get_Queue_History(stock = 'ذوب', start_date = '1400-09-20', end_date='1400-10-20', show_per_capita = True,\
                                     show_weekday = False, double_date = False, show_progress = True)
df_bq_sq_val

Progress :  100.0 %


Day_UL Day_LL         Value      Time    BQ_Value SQ_Value  \
J-Date                                                                  
1400-09-20   3321   3669  110826996114  12:29:58           0        0   
1400-10-05   3443   3805  270344068711  12:29:59           0        0   
1400-10-06   3438   3798  835269163218  12:30:00  3384910530        0   
1400-10-07   3592   3970  268555627554  12:29:59           0        0   
1400-10-08   3599   3977  190918050184  12:29:59           0        0   
1400-10-11   3551   3923  171712691914  12:30:00           0        0   

                 BQPC SQPC  
J-Date                      
1400-09-20          0    0  
1400-10-05          0    0  
1400-10-06  102573046    0  
1400-10-07          0    0  
1400-10-08          0    0  
1400-10-11          0    0

---------------------------------------------
## __3. Reading Market Indices Historical Data__
<font color='green'>__شاخص های بازار بورس تهران__</font>

Market indices play an important role in performace evaluation and developing some investment or trading strategies. For example, most people like to calculate a portfolio or stock's Beta coefficient with respect to overall market index. In this section we will introduce finpy-tse module's functions, you could use to access to Tehran Stock Exchange indices.

### __3.1. Get_CWI_History() Function__
<font color='green'>__Access to Cap-Weighted Index (CWI) Historical Data (= شاخص کل بورس)__</font>

You can use the cap-weighted index (CWI) to check if a portfolio or stock's is ahead of the market or behind it, in term of performance. It can also be used in CAPM and other factor models like Fama-French and Carhart as market factor of the model. The function input parameters are: 


| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low, Close and Volume__ as columns. Remember, Date and Weekday columns can be eliminated by using False values for show_weekday and double_date input parameters.

In [3]:
df_cwi_history = fpy.Get_CWI_History(start_date = '1395-01-01', end_date='1400-12-29',\
                                     ignore_date = False, show_weekday = True, double_date = True)
df_cwi_history

Date    Weekday       Open       High        Low      Close  \
J-Date                                                                         
1395-01-07 2016-03-26   Saturday    80133.0    81200.0    80133.0    81200.3   
1395-01-08 2016-03-27     Sunday    81418.0    81448.0    81255.0    81261.0   
1395-01-09 2016-03-28     Monday    81075.0    81075.0    80936.0    80935.7   
1395-01-10 2016-03-29    Tuesday    80843.0    80890.0    80561.0    80561.3   
1395-01-11 2016-03-30  Wednesday    80775.0    81499.0    80775.0    81480.4   
...               ...        ...        ...        ...        ...        ...   
1400-09-14 2021-12-05     Sunday  1335870.0  1343250.0  1335870.0  1343202.7   
1400-09-15 2021-12-06     Monday  1343510.0  1345570.0  1342290.0  1342353.6   
1400-09-16 2021-12-07    Tuesday  1341580.0  1341580.0  1337510.0  1338016.0   
1400-09-17 2021-12-08  Wednesday  1338260.0  1349360.0  1338260.0  1349389.3   
1400-09-20 2021-12-11   Saturday  1350120.0  1350440.0  1344440.0  1344441.6   

                Volume  
J-Date                  
1395-01-07   785545121  
1395-01-08  1006304416  
1395-01-09   677383208  
1395-01-10  1491153697  
1395-01-11   947224063  
...                ...  
1400-09-14  2778352452  
1400-09-15  2755443049  
1400-09-16  2343601361  
1400-09-17  4037895607  
1400-09-20  4801350997  

[1379 rows x 7 columns]

### __3.2. Get_EWI_History() Function__
<font color='green'>__Access to Equal-Weighted Index (EWI) Historical Data (= شاخص کل هم وزن بورس)__</font>

You can use the equal-weighted index (EWI) to check if a portfolio or stock is ahead of the market or behind it, in term of performance. It can also be used to understand if small-cap stocks performance is better than large-cap stocks, or not.


| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low, Close and Volume__ as columns. Remember, Date and Weekday columns can be eliminated by using False values for show_weekday and double_date input parameters.

In [4]:
df_ewi_history = fpy.Get_EWI_History(start_date = '1395-01-01', end_date='1400-12-29',\
                                     ignore_date = False, show_weekday = False, double_date = False)
df_ewi_history

Open      High       Low     Close      Volume
J-Date                                                        
1395-01-07   13347.0   13450.0   13334.0   13450.1   785545121
1395-01-08   13468.0   13484.0   13461.0   13462.3  1006304416
1395-01-09   13458.0   13474.0   13455.0   13473.0   677383208
1395-01-10   13481.0   13484.0   13467.0   13470.0  1491153697
1395-01-11   13498.0   13600.0   13498.0   13598.9   947224063
...              ...       ...       ...       ...         ...
1400-09-14  362829.0  363367.0  362596.0  363342.0  2778352452
1400-09-15  363586.0  364403.0  363586.0  363776.0  2755443049
1400-09-16  363606.0  363606.0  361138.0  361141.0  2343601361
1400-09-17  361196.0  363586.0  361196.0  363590.0  4037919484
1400-09-20  363785.0  363901.0  361495.0  361495.0  4801350997

[1379 rows x 5 columns]

### __3.3. Get_CWPI_History() Function__
<font color='green'>__Access to Cap-Weighted Price Index (CWPI) Historical Data (= شاخص قیمت وزنی-ارزشی بورس)__</font>

Cap-Weighted Price Index (CWPI) is very similar to Cap-Weighted Index (CWI), except that in CWPI, dividents has no effect in index calculations and there is no adjustment for price changes, resulting from DPS. Please note that the historical data for CWPI index, in the form of OHLCV, is available from 1393-12-04, but adj-close historical data is available from 1387-09-14. The input parameters are:

| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __just_adj_close__ | True/Flase | if True, fucntion returns only adj-close data |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low, Close, Adj-Close and Volume__ as its columns. Remember, __Date__ and __Weekday__ columns can be eliminated by assigning False values for __show_weekday__ and __double_date__ input parameters. You can also use the following setting to get only the adj-close history of CWPI index, which is a highly recommended practice:
```python  
just_adj_close = True 
```

In [26]:
cwpi_history = fpy.Get_CWPI_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                    just_adj_close = False, show_weekday = True, double_date = True)
cwpi_history

Date    Weekday      Open      High       Low     Close  \
J-Date                                                                     
1395-01-07 2016-03-26   Saturday   30479.0   30884.0   30479.0   30884.0   
1395-01-08 2016-03-27     Sunday   30966.0   30979.0   30905.0   30907.0   
1395-01-09 2016-03-28     Monday   30835.0   30835.0   30782.0   30782.0   
1395-01-10 2016-03-29    Tuesday   30747.0   30765.0   30640.0   30640.0   
1395-01-11 2016-03-30  Wednesday   30721.0   30997.0   30721.0   30990.0   
...               ...        ...       ...       ...       ...       ...   
1400-09-15 2021-12-06     Monday  333124.0  333634.0  332820.0  332832.0   
1400-09-16 2021-12-07    Tuesday  332645.0  332645.0  331637.0  331753.0   
1400-09-17 2021-12-08  Wednesday  331822.0  334574.0  331822.0  334574.0   
1400-09-20 2021-12-11   Saturday  334761.0  334842.0  333354.0  333354.0   
1400-09-21 2021-12-12     Sunday  332882.0  332882.0  327469.0  327481.0   

            Adj Close      Volume  
J-Date                             
1395-01-07    30884.2   785545121  
1395-01-08    30906.2  1006304416  
1395-01-09    30782.5   677383208  
1395-01-10    30640.1  1491153697  
1395-01-11    30989.7   947224063  
...               ...         ...  
1400-09-15   332835.0  2755443049  
1400-09-16   331760.0  2343601361  
1400-09-17   334580.0  4037895607  
1400-09-20   333353.0  4801350997  
1400-09-21   327452.0  3811331709  

[1380 rows x 8 columns]

### __3.4. Get_EWPI_History() Function__
<font color='green'>__Access to Equal-Weighted Price Index (EWPI) Historical Data (= شاخص قیمت هم وزن بورس)__</font>

Equal-Weighted Price Index (EWPI) is very similar to Equal-Weighted Index (EWI), except that in EWPI, dividents has no effect in index calculations and there is no adjustment for price changes, resulting from DPS. Please note that the historical data for EWPI index, in the form of OHLCV, is available from 1393-12-05, but adj-close historical data is available from 1392-12-28. The input parameters are:

| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __just_adj_close__ | True/Flase | if True, fucntion returns only adj-close data |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low, Close, Adj-Close and Volume__ as its columns. Remember, __Date__ and __Weekday__ columns can be eliminated by assigning False values for __show_weekday__ and __double_date__ input parameters. You can also use the following setting to get only the adj-close history of EWPI index, which is a highly recommended practice:
```python  
just_adj_close = True 
```

In [33]:
ewpi_history = fpy.Get_EWPI_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = True, \
                                    just_adj_close = False, show_weekday = True, double_date = True)
ewpi_history

Date    Weekday      Open      High       Low     Close  \
J-Date                                                                     
1393-12-05 2015-02-24    Tuesday    9198.0    9212.0    9182.0    9182.0   
1393-12-06 2015-02-25  Wednesday    9183.0    9183.0    9109.0    9109.0   
1393-12-09 2015-02-28   Saturday    9112.0    9117.0    9090.0    9090.0   
1393-12-10 2015-03-01     Sunday    9014.0    9030.0    9009.0    9030.0   
1393-12-11 2015-03-02     Monday    9034.0    9056.0    9034.0    9044.0   
...               ...        ...       ...       ...       ...       ...   
1400-09-15 2021-12-06     Monday  228783.0  229298.0  228783.0  228898.0   
1400-09-16 2021-12-07    Tuesday  228796.0  228796.0  227243.0  227243.0   
1400-09-17 2021-12-08  Wednesday  227280.0  228784.0  227280.0  228783.0   
1400-09-20 2021-12-11   Saturday  228908.0  228982.0  227468.0  227468.0   
1400-09-21 2021-12-12     Sunday  227221.0  227221.0  223766.0  223768.0   

            Adj Close      Volume  
J-Date                             
1393-12-05     9115.9   278521094  
1393-12-06     9033.8   271188153  
1393-12-09     9014.7   257849083  
1393-12-10     9030.4   475683577  
1393-12-11     9043.5   717632564  
...               ...         ...  
1400-09-15   228903.0  2755443049  
1400-09-16   227245.0  2343601361  
1400-09-17   228786.0  4037919484  
1400-09-20   227468.0  4801350997  
1400-09-21   223685.0  3811331709  

[1639 rows x 8 columns]

### __3.5. Get_FFI_History() Function__
<font color='green'>__Access to Free-Float Index (FFI) Historical Data (= TEFIX = شاخص آزاد شناور بورس)__</font>

Free-Float Index (FFI) is very similar to Cap-Weighted Index (CWI), except that in FFI, instead of using all released shares for a stock, we only use number of free float shares to calculate market cap. Please note that the historical data for FFI index, in the form of OHLCV, is available from 1388-04-09, but adj-close historical data is available from 1388-04-08. The input parameters are:

| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __just_adj_close__ | True/Flase | if True, fucntion returns only adj-close data |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low, Close, Adj-Close and Volume__ as its columns. Remember, __Date__ and __Weekday__ columns can be eliminated by assigning False values for __show_weekday__ and __double_date__ input parameters. You can also use the following setting to get only the adj-close history of FFI index, which is a highly recommended practice:
```python  
just_adj_close = True 
```

In [38]:
ffi_history = fpy.Get_FFI_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                  just_adj_close = False, show_weekday = True, double_date = True)
ffi_history

Date    Weekday       Open       High        Low      Close  \
J-Date                                                                         
1395-01-07 2016-03-26   Saturday    92349.0    93869.0    92349.0    93869.0   
1395-01-08 2016-03-27     Sunday    94166.0    94204.0    93830.0    93830.0   
1395-01-09 2016-03-28     Monday    93424.0    93424.0    93262.0    93262.0   
1395-01-10 2016-03-29    Tuesday    93025.0    93052.0    92664.0    92676.0   
1395-01-11 2016-03-30  Wednesday    93026.0    94149.0    93026.0    94147.0   
...               ...        ...        ...        ...        ...        ...   
1400-09-15 2021-12-06     Monday  1713690.0  1716200.0  1709360.0  1709720.0   
1400-09-16 2021-12-07    Tuesday  1707570.0  1707570.0  1702530.0  1705070.0   
1400-09-17 2021-12-08  Wednesday  1706110.0  1724680.0  1706110.0  1724680.0   
1400-09-20 2021-12-11   Saturday  1726490.0  1727160.0  1719790.0  1719790.0   
1400-09-21 2021-12-12     Sunday  1716720.0  1716720.0  1685770.0  1685810.0   

            Adj Close      Volume  
J-Date                             
1395-01-07    93869.0   785545121  
1395-01-08    93830.5  1006304416  
1395-01-09    93262.5   677383208  
1395-01-10    92676.3  1491153697  
1395-01-11    94147.3   947224063  
...               ...         ...  
1400-09-15  1709740.0  2755443049  
1400-09-16  1705090.0  2343601361  
1400-09-17  1724680.0  4037895607  
1400-09-20  1719810.0  4801350997  
1400-09-21  1685800.0  3811331709  

[1380 rows x 8 columns]

### __3.6. Get_MKT1I_History() Function__
<font color='green'>__Access to First Market Index (MKT1I) Historical Data (= شاخص بازار اول بورس)__</font>

In Tehran Stock Exchange's Bourse market, stocks has been divided into two sub-markets, first market and second market, based on their profitiblity, free float, liquidity and some other parameters. Each market has its own cap-weighted index. This function gets the first market index (MKT1I) data for you. Please note that the historical data for MKT1I index, in the form of OHLCV, is available from 1387-09-16, but adj-close historical data is available from 1387-09-14. The input parameters are:

| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __just_adj_close__ | True/Flase | if True, fucntion returns only adj-close data |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low, Close, Adj-Close and Volume__ as its columns. Remember, __Date__ and __Weekday__ columns can be eliminated by assigning False values for __show_weekday__ and __double_date__ input parameters. You can also use the following setting to get only the adj-close history of MKT1I index, which is a highly recommended practice:
```python  
just_adj_close = True 
```

In [20]:
mkt1i_history = fpy.Get_MKT1I_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                      just_adj_close = False, show_weekday = True, double_date = True)
mkt1i_history

Date    Weekday       Open       High        Low      Close  \
J-Date                                                                         
1395-01-07 2016-03-26   Saturday    56906.0    57896.0    56906.0    57897.0   
1395-01-08 2016-03-27     Sunday    58114.0    58145.0    57926.0    57932.0   
1395-01-09 2016-03-28     Monday    57754.0    57754.0    57636.0    57636.0   
1395-01-10 2016-03-29    Tuesday    57533.0    57533.0    57190.0    57192.0   
1395-01-11 2016-03-30  Wednesday    57302.0    57948.0    57302.0    57926.0   
...               ...        ...        ...        ...        ...        ...   
1400-09-15 2021-12-06     Monday  1013570.0  1015120.0  1011050.0  1011240.0   
1400-09-16 2021-12-07    Tuesday  1010390.0  1010390.0  1006430.0  1006970.0   
1400-09-17 2021-12-08  Wednesday  1007340.0  1018870.0  1007340.0  1018870.0   
1400-09-20 2021-12-11   Saturday  1019530.0  1019780.0  1014070.0  1014080.0   
1400-09-21 2021-12-12     Sunday   101176.0   101176.0   993396.0   993440.0   

            Adj Close      Volume  
J-Date                             
1395-01-07    57896.6   627397539  
1395-01-08    57932.0   572942045  
1395-01-09    57635.6   440061076  
1395-01-10    57191.5  1027589626  
1395-01-11    57926.3   676272786  
...               ...         ...  
1400-09-15  1011180.0  1425895564  
1400-09-16  1007020.0  1197622847  
1400-09-17  1018880.0  2671946888  
1400-09-20  1014070.0  3278619780  
1400-09-21   993444.0  2439387339  

[1380 rows x 8 columns]

### __3.7. Get_MKT2I_History() Function__
<font color='green'>__Access to Second Market Index (MKT2I) Historical Data (= شاخص بازار دوم بورس)__</font>

In Tehran Stock Exchange's Bourse market, stocks has been divided into two sub-markets, first market and second market, based on their profitiblity, free float, liquidity and some other parameters. Each market has its own cap-weighted index. This function gets the second market index (MKT2I) data for you. Please note that the historical data for MKT2I index, in the form of OHLCV, is available from 1387-09-16, but adj-close historical data is available from 1387-09-14. The input parameters are:

| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __just_adj_close__ | True/Flase | if True, fucntion returns only adj-close data |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low, Close, Adj-Close and Volume__ as its columns. Remember, __Date__ and __Weekday__ columns can be eliminated by assigning False values for __show_weekday__ and __double_date__ input parameters. You can also use the following setting to get only the adj-close history of MKT2I index, which is a highly recommended practice:
```python  
just_adj_close = True 
```

In [23]:
mkt2i_history = fpy.Get_MKT2I_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                      just_adj_close = False, show_weekday = True, double_date = True)
mkt2i_history

Date    Weekday       Open       High        Low      Close  \
J-Date                                                                         
1395-01-07 2016-03-26   Saturday   170302.0   171185.0   170302.0   171185.0   
1395-01-08 2016-03-27     Sunday   171268.0   171373.0   171268.0   171359.0   
1395-01-09 2016-03-28     Monday   171235.0   171235.0   171046.0   171057.0   
1395-01-10 2016-03-29    Tuesday   171081.0   171466.0   171081.0   171322.0   
1395-01-11 2016-03-30  Wednesday   172018.0   172787.0   172018.0   172787.0   
...               ...        ...        ...        ...        ...        ...   
1400-09-15 2021-12-06     Monday  2614510.0  2618570.0  2614440.0  2616770.0   
1400-09-16 2021-12-07    Tuesday  2616190.0  2616190.0  2610660.0  2611390.0   
1400-09-17 2021-12-08  Wednesday  2611460.0  2623100.0  2609710.0  2623100.0   
1400-09-20 2021-12-11   Saturday  2624280.0  2625370.0  2616950.0  2616950.0   
1400-09-21 2021-12-12     Sunday  2615990.0  2615990.0  2579490.0  2579580.0   

            Adj Close      Volume  
J-Date                             
1395-01-07   171185.0   158147582  
1395-01-08   171359.0   433362371  
1395-01-09   171057.0   237322132  
1395-01-10   171322.0   463564071  
1395-01-11   172787.0   270951277  
...               ...         ...  
1400-09-15  2617000.0  1329547485  
1400-09-16  2611370.0  1145978514  
1400-09-17  2623170.0  1365972596  
1400-09-20  2616940.0  1522731217  
1400-09-21  2579630.0  1371944370  

[1380 rows x 8 columns]

### __3.8. Get_INDI_History() Function__
<font color='green'>__Access to Industry Index (INDI) Historical Data (= شاخص صنعت بورس)__</font>

This is a cap-weighted index for just those stocks that are industrial companies. This function gets the industry index (INDI) data for you. Please note that the historical data for MKT2I index, in the form of OHLCV, is available from 1387-09-15, but adj-close historical data is available from 1387-09-14. The input parameters are:

| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __just_adj_close__ | True/Flase | if True, fucntion returns only adj-close data |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low, Close, Adj-Close and Volume__ as its columns. Remember, __Date__ and __Weekday__ columns can be eliminated by assigning False values for __show_weekday__ and __double_date__ input parameters. You can also use the following setting to get only the adj-close history of INDI index, which is a highly recommended practice:
```python  
just_adj_close = True 
```

In [29]:
indi_history = fpy.Get_INDI_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                    just_adj_close = False, show_weekday = True, double_date = True)
indi_history

Date    Weekday       Open       High        Low      Close  \
J-Date                                                                         
1395-01-07 2016-03-26   Saturday    66988.0    67547.0    66988.0    67547.0   
1395-01-08 2016-03-27     Sunday    67595.0    67607.0    67493.0    67493.0   
1395-01-09 2016-03-28     Monday    67276.0    67314.0    67209.0    67209.0   
1395-01-10 2016-03-29    Tuesday    67166.0    67166.0    66874.0    66875.0   
1395-01-11 2016-03-30  Wednesday    67102.0    67589.0    67102.0    67589.0   
...               ...        ...        ...        ...        ...        ...   
1400-09-15 2021-12-06     Monday  1245020.0  1247150.0  1243940.0  1243940.0   
1400-09-16 2021-12-07    Tuesday  1242280.0  1242280.0  1239530.0  1240190.0   
1400-09-17 2021-12-08  Wednesday  1240570.0  1250790.0  1240570.0  1250790.0   
1400-09-20 2021-12-11   Saturday  1251260.0  1251260.0  1245610.0  1245610.0   
1400-09-21 2021-12-12     Sunday  1240790.0  1240790.0  1222650.0  1222710.0   

            Adj Close      Volume  
J-Date                             
1395-01-07    67547.1   571954728  
1395-01-08    67492.9   529072524  
1395-01-09    67209.4   397361929  
1395-01-10    66874.7  1151654864  
1395-01-11    67588.8   681107453  
...               ...         ...  
1400-09-15  1243950.0  1837345052  
1400-09-16  1240220.0  1537452505  
1400-09-17  1250820.0  3148015494  
1400-09-20  1245600.0  3726868638  
1400-09-21  1222720.0  3019503110  

[1380 rows x 8 columns]

### __3.9. Get_LCI30_History() Function__
<font color='green'>__Access to Top 30 Large-Cap Index (LCI30) Historical Data (= شاخص 30 شرکت بزرگ بورس)__</font>

This is a cap-weighted index for top 30 large-cap (LCI30) stocks in Bourse market. This function gets the 30 large-cap index (LCI30) data for you. Please note that the historical data for 30LCI index, in the form of OHLCV and adj-close, both are available from 1389-05-24. The input parameters are:

| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __just_adj_close__ | True/Flase | if True, fucntion returns only adj-close data |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low, Close, Adj-Close and Volume__ as its columns. Remember, __Date__ and __Weekday__ columns can be eliminated by assigning False values for __show_weekday__ and __double_date__ input parameters. You can also use the following setting to get only the adj-close history of LCI30 index, which is a highly recommended practice:
```python  
just_adj_close = True 
```

In [35]:
lci30_history = fpy.Get_LCI30_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                      just_adj_close = False, show_weekday = True, double_date = True)
lci30_history

Date    Weekday     Open     High      Low    Close  \
J-Date                                                                 
1395-01-07 2016-03-26   Saturday   3379.0   3454.0   3379.0   3454.0   
1395-01-08 2016-03-27     Sunday   3465.0   3475.0   3455.0   3455.0   
1395-01-09 2016-03-28     Monday   3444.0   3444.0   3431.0   3431.0   
1395-01-10 2016-03-29    Tuesday   3421.0   3423.0   3401.0   3401.0   
1395-01-11 2016-03-30  Wednesday   3404.0   3452.0   3404.0   3450.0   
...               ...        ...      ...      ...      ...      ...   
1400-09-15 2021-12-06     Monday  72006.0  72087.0  71633.0  71682.0   
1400-09-16 2021-12-07    Tuesday  71667.0  71667.0  71395.0  71525.0   
1400-09-17 2021-12-08  Wednesday  71530.0  72551.0  71530.0  72545.0   
1400-09-20 2021-12-11   Saturday  72547.0  72632.0  72330.0  72330.0   
1400-09-21 2021-12-12     Sunday  72298.0  72298.0  70637.0  70639.0   

            Adj Close      Volume  
J-Date                             
1395-01-07     3454.2   290228382  
1395-01-08     3455.4   336982051  
1395-01-09     3430.6   255295732  
1395-01-10     3400.7   411464948  
1395-01-11     3450.4   323150383  
...               ...         ...  
1400-09-15    71679.0   513710731  
1400-09-16    71526.5   490790033  
1400-09-17    72546.4  1311043374  
1400-09-20    72331.2  1734330497  
1400-09-21    70638.1  1263718413  

[1380 rows x 8 columns]

### __3.10. Get_ACT50_History() Function__
<font color='green'>__Access to Top 50 Active Index (ACT50) Historical Data (= شاخص 50 شرکت فعال بورس)__</font>

This is a cap-weighted index for top 50 active (ACT50) stocks in Bourse market. This function gets the 50 most active stocks index (ACT50) data for you. Please note that the historical data for ACT50 index, in the form of OHLCV, is available from 1387-09-16, but adj-close historical data is available from 1387-09-15. The input parameters are:

| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __just_adj_close__ | True/Flase | if True, fucntion returns only adj-close data |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low, Close, Adj-Close and Volume__ as its columns. Remember, __Date__ and __Weekday__ columns can be eliminated by assigning False values for __show_weekday__ and __double_date__ input parameters. You can also use the following setting to get only the adj-close history of ACT50 index, which is a highly recommended practice:
```python  
just_adj_close = True 
```

In [42]:
act50_history = fpy.Get_ACT50_History(start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                      just_adj_close = False, show_weekday = True, double_date = True)
act50_history

Date    Weekday     Open     High      Low    Close  \
J-Date                                                                 
1395-01-07 2016-03-26   Saturday   3256.0   3301.0   3256.0   3301.0   
1395-01-08 2016-03-27     Sunday   3311.0   3311.0   3305.0   3305.0   
1395-01-09 2016-03-28     Monday   3293.0   3296.0   3290.0   3290.0   
1395-01-10 2016-03-29    Tuesday   3287.0   3287.0   3271.0   3271.0   
1395-01-11 2016-03-30  Wednesday   3283.0   3311.0   3283.0   3311.0   
...               ...        ...      ...      ...      ...      ...   
1400-09-16 2021-12-07    Tuesday  52726.0  52726.0  52594.0  52633.0   
1400-09-17 2021-12-08  Wednesday  52647.0  53126.0  52644.0  53126.0   
1400-09-20 2021-12-11   Saturday  53148.0  53148.0  52933.0  52933.0   
1400-09-21 2021-12-12     Sunday  52717.0  52717.0  51926.0  51928.0   
1400-09-22 2021-12-13     Monday  51871.0  51871.0  50871.0  50871.0   

            Adj Close      Volume  
J-Date                             
1395-01-07     3300.7   631986992  
1395-01-08     3304.8   775084867  
1395-01-09     3289.8   470238679  
1395-01-10     3270.6  1127454520  
1395-01-11     3310.8   644347187  
...               ...         ...  
1400-09-16    52635.0   821511900  
1400-09-17    53126.7  2287125171  
1400-09-20    52932.7  2965642747  
1400-09-21    51928.8  2212036794  
1400-09-22    50872.3  2156107973  

[1381 rows x 8 columns]

### __3.11. Get_SectorIndex_History() Function__
<font color='green'>__Access to a Sector Index Historical Data (= شاخص گروه صنعت در بازار بورس)__</font>

Each sector has its own cap-weighted index. The universe of each sector index is that sector's stocks in the Bourse (not Fara-Bourse) market. You can access to each sector cap-weighted index using this function and sector name. Sector names are listed in the following table:

| __Sector Name (EN)__ | __Sector Name (FA)__ | __Sector Name (EN)__ | __Sector Name (FA)__ |  __Sector Name (EN)__ | __Sector Name (FA)__ |
| :-: | :-: | :-: |:-: |:-: |:-: |
|__Agriculture__| __زراعت__|__Coal__| __ذغال سنگ__ |__Metal Ore__| __کانه فلزی__|__Other Mines__| __سایر معادن__ |
|__Textiles__| __منسوجات__|__Leather Products__| __محصولات چرمی__ |__Wood Products__| __محصولات چوبی__|__Paper Products__| __محصولات کاغذی__ |
|__Publishing__| __انتشار و چاپ__|__Petroleum Products__| __فرآورده های نفتی__ |__Rubber__| __لاستیک__|__Basic Metals__| __فلزات اساسی__ |
|__Metal Products__| __محصولات فلزی__|__Machinery__| __ماشین آلات__ |__Suger__| __قند و شکر__|__Multi-Discipline__| __چند رشته ای__ |
|__Electrical Devices__| __دستگاه های برقی__|__Communication Tools__| __وسایل ارتباطی__ |__Car__| __خودرو__ |__Insurance__| __بیمه و بازنشستگی__ |
|__Drugs__| __مواد دارویی__|__Chemicals__| __شیمیایی__ |__Retail__| __خرده فروشی به جز وسایل نقلیه__|__Ceramic Tiles__| __كاشی و سرامیک__ |
|__Cement__| __سیمان__|__Non-Metalic Ore__| __کانی غیر فلزی__ |__Computer__| __رایانه__|__IT__| __اطلاعات و ارتباطات__ |
|__Investment__| __سرمایه گذاری__|__Banks__| __بانک__ |__Other Financials__| __سایر مالی__|__Transportation__| __حمل و نقل__ |
|__Radio__| __رادیویی__|__Finance__| __مالی__ |__Management of Financial Markets__| __اداره بازارهای مالی__|__Building__| __انبوه سازی__ |
|__Oil Extraction__| __استخراج نفت__||||

The input parameters are:

| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __sector__ | Sector name in Farsi | For quickest response, enter the sector name just like the above table, otherwise the response will be 1~2 second slower |
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date (included, if it is a trading day) |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date (included, if it is a trading day) |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __just_adj_close__ | True/Flase | if True, fucntion returns only adj-close data |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low, Close, Adj-Close and Volume__ as its columns. Remember, __Date__ and __Weekday__ columns can be eliminated by assigning False values for __show_weekday__ and __double_date__ input parameters. You can also use the following setting to get only the adj-close history ofthe sector index, which is a highly recommended practice:
```python  
just_adj_close = True 
```

In [115]:
df_sector_index_history = fpy.Get_SectorIndex_History(sector = 'چند رشته ای', start_date='1395-01-01', end_date='1400-12-29', ignore_date = False, \
                                                      just_adj_close = True, show_weekday = True, double_date = True)
df_sector_index_history

Date    Weekday  Adj Close
J-Date                                     
1395-01-07 2016-03-26   Saturday     7696.8
1395-01-08 2016-03-27     Sunday     7661.2
1395-01-09 2016-03-28     Monday     7583.5
1395-01-10 2016-03-29    Tuesday     7525.5
1395-01-11 2016-03-30  Wednesday     7511.7
...               ...        ...        ...
1400-09-23 2021-12-14    Tuesday   164324.0
1400-09-24 2021-12-15  Wednesday   163806.0
1400-09-27 2021-12-18   Saturday   164856.0
1400-09-28 2021-12-19     Sunday   162465.0
1400-09-29 2021-12-20     Monday   163565.0

[1387 rows x 3 columns]

---------------------------------------------
## __4. Reading Real-Time Market Data__

ُSometimes, traders make their decisions based on the market real time data, when the market is open. For example you may have an strategy based on the real-time orderbook data

The function will return two dataframes, the first one is the market watch real-time data and the second is the order book real-time data. The first dataframe have the following columns:

| __Column Name__ |__Description (EN)__ | __Description (FA)__ | 
| :- | :- | -: | 
| __Ticker (index)__ |Ticker| نماد |
| __Trade Type__ |Trade Type| نوع معامله: تابلو، بلوکی، عمده، جبرانی |
| __Time__ |Time| زمان مربوط به اطلاعات نمایش داده شده یک نماد |
| __Open__ |Day Open Price| قیمت باز شدن روز جاری |
| __High__ |Day High Price| بالاترین قیمت رخ داده در روز جاری تا آن لحظه |
| __Low__ |Day Low Price| پایین ترین قیمت رخ داده در روز جاری تا آن لحظه |
| __Close__ |Day Close Price| آخرین قیمت معامله در آن لحظه |
| __Final__ |Day Final Price| قیمت پایانی در آن لحظه |
| __Close(%)__ |Day Close % Change| درصد تغییر آخرین قیمت معامله در آن لحظه نسبت به قیمت پایانی دیروز |
| __Final(%)__ |Day Final % Change| درصد تغییر قیمت پایانی در آن لحظه نسبت به قیمت پایای دیروز |
| __Day_UL__ |Day Upper Limit Price| بالاترین قیمت مجاز روز جاری |
| __Day_LL__ |Day Lower Limit Price| پایین ترین قیمت مجاز روز جاری |
| __Value__ |Value of Executed Trades| ارزش معاملات در روز جاری تا آن لحظه |
| __BQ-Value__ |Buy Queue Value| ارزش صف خرید در نماد مورد نظر در آن لحظه |
| __SQ-Value__ |Sell Queue Value| ارزش صف فروش در نماد مورد نظر در آن لحظه |
| __BQPC-avg__ |Buy Queue Per-Capita (Average)| ارزش سرانه افرادی که در آن لحظه در صف خرید نماد مورد نظر هستند |
| __SQPC-avg__ |Sell Queue Per-Capita (Average)| ارزش سرانه افرادی که در آن لحظه در صف فروش نماد مورد نظر هستند |
| __Volume__ |Volume of Executed Trades| حجم معاملات در نماد مورد نظر تا آن لحظه |
| __Vol_Buy_R__ |Retail Traders Buy Volume| حجم خرید افراد حقیقی در نماد مورد نظر تا آن لحظه |
| __Vol_Buy_I__ |Institutioanl Traders Buy Volume| حجم خرید افراد حقوقی در نماد مورد نظر تا آن لحظه |
| __Vol_Sell_R__ |Retail Traders Sell Volume| حجم فروش افراد حقیقی در نماد مورد نظر تا آن لحظه |
| __Vol_Sell_I__ |Institutional Traders Sell Volume| حجم فروش افراد حقوقی در نماد مورد نظر تا آن لحظه |
| __No__ |Number of Executed Trades| تعداد معاملات انجام شده در نماد مورد نظر تا آن لحظه |
| __No_Buy_R__ |Number of Buyer Retail Traders| تعداد خریداران حقیقی در نماد مورد نظر تا آن لحظه |
| __No_Sell_R__ |Number of Seller Retail Traders| تعداد فروششندگان حقیقی در نماد مورد نظر تا آن لحظه |
| __No_Buy_I__ |Number of Buyer Institutional Retail Traders| تعداد خریداران حقوقی در نماد مورد نظر تا آن لحظه |
| __No_Sell_I__ |Number of Seller Institutional Traders| تعداد فروشندگان حقوقی در نماد مورد نظر تا آن لحظه |
| __Name__ |Firm Farsi Full Name| نام فارسی شرکت مربوط به نماد مورد نظر |
| __Market__ |Category and Market of the Ticker| بازار نماد مورد نظر: بورس، فرابورس، پایه، حق تقتدم بورس، حق تقدم فرابورس، حق تقدم پایه و صندوق های قابل معامله |
| __Sector__ |Sector/Industry of the ticker| گروه صنعت نماد مورد نظر |
| __Share-No__ |Number of Released Shares of the Ticker| تعداد سهام نماد مورد نظر |
| __Base-Vol__ |Base Volume of the Ticker| حجم مبنای نماد مورد نظر |
| __Market Cap__ |Market Cap of the Ticker (based on Final Price)| ارزش بازار نماد مورد نظر در آن لحظه - محاسبه شده بر اساس قیمت پایانی|
| __EPS__ | TTM EPS (TSE Data) | نماد مورد نظر به صورت میانگین دوازده ماهه متحرک EPS|
| __Download__ |Download Jalali Date and Time| تاریخ شمسی و زمان دانلود اطلاعات |

The second dataframe is for real-time order book at that moment, with the following columns:

| __Column Name__ |__Description (EN)__ | __Description (FA)__ | 
| :- | :- | -: | 
| __Ticker (index)__ |Ticker| نماد |
| __Day_UL__ |Day Upper Limit Price| بالاترین قیمت مجاز برای معامله در روز جاری |
| __Day_LL__ |Day Lower Limit Price| پایین ترین قیمت مجاز برای معامله در روز جاری |
| __OB-Depth__ |Order Book Depth (1-5)| عمق دفتر سفارشات (بهترین مظنه در عمق 1) |
| __Sell-No__ |Number of People Willing to Sell at Given Price| تعداد افراد حاضر به فروش در آن مظنه فروش |
| __Sell-Vol__ |Day Open Price| حجم تجمعی فروشندگان در آن مطنه فروش |
| __Sell-Price__ |Offered Sell Price| قیمت پیشنهادی فروشندگان در آن مظنه فروش |
| __Buy-Price__ |Offered Buy Price| قیمت پیشنهادی خریداران در آن مظنه خرید |
| __Buy-Vol__ |Ready Buy Volume at Given Price| حجم تجمعی خریداران در آن مظنه خرید  |
| __Buy-No__ |Number of People Willing to Buy at Given Price| تعداد افراد حاضر به خرید در ان مظنه خرید |
| __Download__ |Download Jalali Date and Time| تاریخ شمسی و زمان دانلود اطلاعات |

Both dataframes will be returned by this function and you can manage them in your desired databases if you will. If you want the dataframes to be saved as excel in your local drive, you need to say this to the function and specofy a path for saving data using the following input parameters:
```python  
save_excel = True
save_path = 'D:/FinPy-TSE Data/MarketWatch'
```

In [52]:
df_market_watch, df_order_book = fpy.Get_MarketWatch(save_excel = True, save_path = 'D:/FinPy-TSE Data/MarketWatch')
df_market_watch.head(5)

Trade Type      Time     Open     High      Low    Close    Final  \
Ticker                                                                     
وسپهر       تابلو  12:29:31   7060.0   7061.0   7060.0   7060.0   7067.0   
شصدف        تابلو  12:29:51  34200.0  34560.0  33501.0  34000.0  33769.0   
فسا         تابلو  13:13:16    918.0    939.0    916.0    916.0    935.0   
فرآور       تابلو  12:29:21  28880.0  28880.0  27200.0  27700.0  27740.0   
سبزوا       تابلو  12:29:29   1859.0   1914.0   1818.0   1836.0   1862.0   

        Close(%)  Final(%)   Day_UL  ...  No_Sell_R  No_Sell_I  \
Ticker                               ...                         
وسپهر      -0.16     -0.06   7424.0  ...     1331.0        3.0   
شصدف       -2.64     -3.30  36669.0  ...      606.0        3.0   
فسا        -2.97     -0.95    972.0  ...       52.0        0.0   
فرآور      -2.94     -2.80  29960.0  ...       92.0        0.0   
سبزوا      -4.03     -2.67   2008.0  ...      169.0        0.0   

                                Name   Market           Sector      Share-No  \
Ticker                                                                         
وسپهر   سرمایه گذاری مالی سپهرصادرات  فرابورس   سرمایه گذاریها  3.000000e+10   
شصدف                  صنعتی دوده فام  فرابورس  محصولات شیمیایی  7.000000e+08   
فسا                     پتروشیمی فسا     پایه  محصولات شیمیایی  3.486000e+09   
فرآور        فرآوری موادمعدنی ایران      بورس      فلزات اساسی  6.600000e+08   
سبزوا               سیمان لار سبزوار  فرابورس  سیمان، آهک و گچ  3.757798e+09   

          Base-Vol    Market Cap     EPS             Download  
Ticker                                                         
وسپهر   12000000.0  2.120100e+14   444.0  1400-10-11 20:50:19  
شصدف      429517.0  2.363830e+13  2137.0  1400-10-11 20:50:19  
فسا     10593221.0  3.259410e+12     2.0  1400-10-11 20:50:19  
فرآور     525578.0  1.830840e+13  1005.0  1400-10-11 20:50:19  
سبزوا    7841088.0  6.997019e+12   289.0  1400-10-11 20:50:19  

[5 rows x 34 columns]

In [54]:
df_order_book.head(5)

Sell-No  Sell-Vol  Sell-Price  Buy-Price  \
Ticker Day_LL Day_UL OB-Depth                                             
آبادا  9730   10750  1               1      4224       10080      10010   
                     2               6       793       10090      10000   
                     3               5      2969       10100       9980   
                     4               5       692       10120       9970   
                     5               2        96       10130       9960   

                               Buy-Vol  Buy-No             Download  
Ticker Day_LL Day_UL OB-Depth                                        
آبادا  9730   10750  1            7216       2  1400-10-11 20:50:19  
                     2           48258       3  1400-10-11 20:50:19  
                     3            1000       1  1400-10-11 20:50:19  
                     4            1466       2  1400-10-11 20:50:19  
                     5           99671       2  1400-10-11 20:50:19

---------------------------------------------
## __5. Collective Data Download__

### __5.1. Build_Market_StockList() Function__
<font color='green'>__Gathering Detailed Information of All Stocks, Available in Tehran Stock Exchange (TSE)__</font>

Every investor or trader needs to know all of the options available for investment or trading in a given market. This function, provides you list of all stocks from Bourse, Fara-Bourse and Payeh market. It would be a good practice to save the output of this function in your local drive, play around with it in Python (or even Excel using Filter function), to choose a list of stocks that meets your initial requirement. We have provided some preliminay option for you.

In [99]:
stock_list = fpy.Build_Market_StockList(bourse = True, farabourse = True, payeh = True, detailed_list = True, show_progress = True, 
                                        save_excel = True, save_csv = False, save_path = 'D:/FinPy-TSE Data/')
stock_list

Progress : 100 % , Done in 20 seconds!
File saved in the specificed directory as:  1400-10-11 detailed_stocklist_bfp.xlsx


Name       Market  \
Ticker                                             
آبادا        تولید نیروی برق آبادان         بورس   
آپ                آسان پرداخت پرشین         بورس   
آسیا                      بیمه آسیا         بورس   
آکنتور             کنتورسازی ایران          بورس   
اپال    فرآوری معدنی اپال کانی پارس         بورس   
...                             ...          ...   
لکما      کارخانجات مخابراتی ایران      پایه زرد   
غیوان                         کیوان    پایه قرمز   
کهرام          تولیدی گرانیت بهسرام  پایه نارنجی   
سایرا                      ایرانیت   پایه نارنجی   
سباقر                  سیمان باقران     پایه زرد   

                               Panel  \
Ticker                                 
آبادا                 بازار دوم بورس   
آپ                    بازار دوم بورس   
آسیا    بازار اول (تابلوی اصلی) بورس   
آکنتور  بازار اول (تابلوی فرعی) بورس   
اپال                  بازار دوم بورس   
...                              ...   
لکما          بازار پایه زرد فرابورس   
غیوان        بازار پایه قرمز فرابورس   
کهرام      بازار پایه نارنجی فرابورس   
سایرا      بازار پایه نارنجی فرابورس   
سباقر         بازار پایه زرد فرابورس   

                                          Sector  \
Ticker                                             
آبادا                 عرضه برق، گاز، بخاروآب گرم   
آپ              رایانه و فعالیت های وابسته به آن   
آسیا    بیمه وصندوق بازنشستگی به جزتامین اجتماعی   
آکنتور          ابزارپزشکی، اپتیکی و اندازه گیری   
اپال                       استخراج کانه های فلزی   
...                                          ...   
لکما              ساخت دستگاه ها و وسایل ارتباطی   
غیوان   محصولات غذایی و آشامیدنی به جز قند و شکر   
کهرام                              کاشی و سرامیک   
سایرا                  سایر محصولات کانی غیرفلزی   
سباقر                            سیمان، آهک و گچ   

                                     Sub-Sector  \
Ticker                                            
آبادا            تولید انتقال و توزیع نیروی برق   
آپ                      مشاوره و تهیه نرم افزار   
آسیا                              بیمه غیرزندگی   
آکنتور     وسایل اندازه گیری غیر از کنترل صنعتی   
اپال                                استخراج آهن   
...                                         ...   
لکما    فرستنده رادیو، تلویزیون و تلفن و تلگراف   
غیوان                تولید نان و محصولات وابسته   
کهرام                تولید گل و سرامیک ساختمانی   
سایرا           تولید محصولات بتنی، سیمانی و گچ   
سباقر                     تولید سیمان، آهک و گچ   

                                         Comment            Name(EN)  \
Ticker                                                                 
آبادا                                          -           Abadan PG   
آپ                                             -  Asan Pardakht Pers   
آسیا                                           -           Asia Bime   
آکنتور                                         -        Iran Counter   
اپال                                           -      Opal Kani Pars   
...                                          ...                 ...   
لکما                               لغو پذیرش شده     Iran Tele Fact.   
غیوان                 لغو پذیرش شده/عدم اظهارنظر              Keivan   
کهرام                              لغو پذیرش شده    Behceram Granite   
سایرا              لغو پذیرش شده/ اظهارنظر مردود              Iranit   
سباقر    قانون احکام دائمی برنامه های توسعه کشور     Bagheran Cement   

       Company Code(12) Ticker(4) Ticker(5)    Ticker(12) Sector Code  \
Ticker                                                                  
آبادا      IRO1NBAB0009      NBAB     NBAB1  IRO1NBAB0001          40   
آپ         IRO1APPE0008      APPE     APPE1  IRO1APPE0001          72   
آسیا       IRO1ASIA0008      ASIA     ASIA1  IRO1ASIA0001          66   
آکنتور     IRO1CONT0002      CONT     CONT1  IRO1CONT0001          33   
اپال       IRO1OPAL0000      OPAL     OPAL1  IRO1OPAL0001          13   
...                 ...       ...       ... 

### __5.2. Build_Price_Panel() Function__
<font color='green'>__Gathering Detailed Information of All Stocks, Available in Tehran Stock Exchange (TSE)__</font>

In stock selection, portfolio management and some other applications, we need to have some sort tabular data, to be able to feed our algorithm. For example, in the following table, you can see adj-close price data of 10 stocks for 1000 days. Columns of this table are tickers (or stock name) with dates as index. Inside the table, adj-close price of each stock for that date is stored. The best practice is to have this table in the form of a Pandas data frame.

| __Date__ | __Stock 1__ | __Stock 2__ | __Stock 3__ |__Stock 4__ |__Stock 5__ | __Stock 6__ | __Stock 7__ | __Stock 8__ |__Stock 9__ |__Stock 10__ |
| :- | :-: | :-: |:-: | :-: | :-: |:-: | :-: |:-: | :-: | :-: | 
| __Date 1__ | $P_{1,1}$ |$P_{2,1}$ |$P_{3,1}$ |$P_{4,1}$ |$P_{5,1}$ |$P_{6,1}$ |$P_{7,1}$ |$P_{8,1}$ |$P_{9,1}$ |$P_{10,1}$ |
| __Date 2__ | $P_{1,2}$ |$P_{2,2}$ |$P_{3,2}$ |$P_{4,2}$ |$P_{5,2}$ |$P_{6,2}$ |$P_{7,2}$ |$P_{8,2}$ |$P_{9,2}$ |$P_{10,2}$ |
| __Date 3__ | $P_{1,3}$ |$P_{2,3}$ |$P_{3,3}$ |$P_{4,3}$ |$P_{5,3}$ |$P_{6,3}$ |$P_{7,3}$ |$P_{8,3}$ |$P_{9,3}$ |$P_{10,3}$ |
| __...__ | ... |... |... |... |... |... |... |... |... |... |
| __Date 500__ | $P_{1,500}$ |$P_{2,500}$ |$P_{3,500}$ |$P_{4,500}$ |$P_{5,500}$ |$P_{6,500}$ |$P_{7,500}$ |$P_{8,500}$ |$P_{9,500}$ |$P_{10,500}$ |
| __Date 501__ | $P_{1,501}$ |$P_{2,501}$ |$P_{3,501}$ |$P_{4,501}$ |$P_{5,501}$ |$P_{6,501}$ |$P_{7,501}$ |$P_{8,501}$ |$P_{9,501}$ |$P_{10,501}$ |
| __Date 502__ | $P_{1,502}$ |$P_{2,502}$ |$P_{3,502}$ |$P_{4,502}$ |$P_{5,502}$ |$P_{6,502}$ |$P_{7,502}$ |$P_{8,502}$ |$P_{9,502}$ |$P_{10,502}$ |
| __...__ | ... |... |... |... |... |... |... |... |... |... |
| __Date 998__ | $P_{1,998}$ |$P_{2,998}$ |$P_{3,998}$ |$P_{4,998}$ |$P_{5,998}$ |$P_{6,998}$ |$P_{7,998}$ |$P_{8,998}$ |$P_{9,998}$ |$P_{10,998}$ |
| __Date 999__ | $P_{1,999}$ |$P_{2,999}$ |$P_{3,999}$ |$P_{4,999}$ |$P_{5,999}$ |$P_{6,999}$ |$P_{7,999}$ |$P_{8,999}$ |$P_{9,999}$ |$P_{10,999}$ |
| __Date 1000__ | $P_{1,1000}$ |$P_{2,1000}$ |$P_{3,1000}$ |$P_{4,1000}$ |$P_{5,1000}$ |$P_{6,1000}$ |$P_{7,1000}$ |$P_{8,1000}$ |$P_{9,1000}$ |$P_{10,1000}$ |

The functions, we will introduce in this section are used to download your desired stocks data, store that data in your local drive and then build tables like the above one. You can choose any other variable for building the above table, for example, Open, High, Low, Close and ..., but the default variable is __Adj Close__.

---------------------------------------------
## __6. Reading USD/Rial Exchange Rate__

Although USD/Rial exchange rate is out of the scope of Tehran Stock Exchange, but as it has important effect on the price of stocks, we decided to include a function in finpy-tse module, to access this data, as well. 

### __6.1. Get_USD-RIAL() Function__
<font color='green'>__Access to USD/RIAL Exchange Rate Historical Data (= نرخ دلار آمریکا در بازار آزاد)__</font>

The function has similar inputs to our other functions. It returns daily price of USD and only supports 10-years of daily history.


| __Parameter__ | __Format__ | __Description__ | 
| :- | :- | :- | 
| __start_date__ | YYYY-MM-DD Jalali Date | returned data start date |
| __end_date__ | YYYY-MM-DD Jalali Date | returned data end date |
| __ignore_date__ | True/Flase | if True, fucntion do not consider start and end date and gets all available history |
| __show_weekday__ | True/Flase | if True, fucntion shows weekdays in a separate column |
| __double_date__ | True/Flase | if True, fucntion shows international date in a separate column, in addition to Jalali date |

The output of the function will be a pandas dataframe with __J-Date__ as index and __Date, Weekday, Open, High, Low and Close__ as columns. Remember, Date and Weekday columns can be eliminated by using False values for show_weekday and double_date input parameters.

In [5]:
df_usd = fpy.Get_USD_RIAL(start_date = '1395-01-01', end_date = '1400-12-29',\
                          ignore_date = False, show_weekday = True, double_date = True)
df_usd

Date    Weekday    Open    High     Low   Close
J-Date                                                          
1395-01-07 2016-03-26   Saturday   34300   34750   34300   34680
1395-01-08 2016-03-27     Sunday   34660   34670   34660   34660
1395-01-10 2016-03-29    Tuesday   34960   35140   34960   35130
1395-01-11 2016-03-30  Wednesday   34890   34890   34820   34820
1395-01-12 2016-03-31   Thursday   34850   34850   34850   34850
...               ...        ...     ...     ...     ...     ...
1400-09-15 2021-12-06     Monday  278720  278720  278080  278240
1400-09-16 2021-12-07    Tuesday  278200  278240  278080  278130
1400-09-17 2021-12-08  Wednesday  278160  278240  277610  277770
1400-09-18 2021-12-09   Thursday  277640  277770  277000  277040
1400-09-20 2021-12-11   Saturday  277160  277160  276820  276950

[1649 rows x 6 columns]